### DISCUSSION TITLE
Learning from in-edges and out-edges

### DISCUSSION THREAD

**Discussion by YoavLotem** [pyg-team/pytorch_geometric#5106] [Q&A]
Hi, 
Thanks for this awesome tool : )

I'm working on task with a directional graph.
I would like to have message passing in a bi-directional way but since in-edges and out-edges have different analytical meaning I would like to have different weight matrices - therefore turning the graph into an undirected graph is not the best option.

Are you aware of any implementation that solves my problem?

I guess it might be possible to construct a heterogenous graph with a "src-dst" edge and a "dst-src" edge, but I was wondering if existing solutions are available.

Thanks in advance!

**Comment by wsad1** [comment]
[GCNConv](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.GCNConv) and several other [layers](https://pytorch-geometric.readthedocs.io/en/latest/notes/cheatsheet.html), have support for `edge_weight`. How about setting the `edge_weight` to based on if its and in or out edge?

**Reply by YoavLotem** [reply]
Hi, Thanks for your response.

I would like to have different weight matrices for the in & out edges.
I guess that weighting them differently but having the same weight matrix wouldn't be as expressive.

Are there any conv layers that have one weight matrix for in edges and another for out edges?
Similarly to how some layers have weight matrices to incorporate edge features (for example TransformerConv).

**Reply by wsad1** [reply]
Ah I see then your looking for something like [RGCNConv](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.RGCNConv), [RGATConv](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.RGATConv) or [FilmConv](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.FiLMConv) which have different weights matrices for each edge type.

**Reply by YoavLotem** [reply]
Thanks!!
Do you know if any of these handle directed graphs this way automatically or I need to define 2 types of edges "src-dst" and "dst-src"?

**Reply by wsad1** [reply]
You need to define two edge types if you want different weight matrices.

**Reply by YoavLotem** [reply]
Got it. 
Thanks!
